src/Data/HashTable/ST/Basic.hs

```diff
@@ -460,56 +483,46 @@ growTable ht@(HashTable sz _ _ _ _) = do
 
 
 ------------------------------------------------------------------------------
--- Helper data structure for delete'
+-- Helper data structure for findSafeSlots
 data Slot = Slot {
       _slot       :: {-# UNPACK #-} !Int
-    , _wasDeleted :: {-# UNPACK #-} !Int  -- we use Int because Bool won't
-                                          -- unpack
     }
   deriving (Show)
 
 
 ------------------------------------------------------------------------------
 instance Monoid Slot where
-    mempty = Slot maxBound 0
-    (Slot x1 b1) `mappend` (Slot x2 b2) =
-        if x1 == maxBound then Slot x2 b2 else Slot x1 b1
+    mempty = Slot maxBound
+    (Slot x1) `mappend` (Slot x2) =
+        if x1 == maxBound then Slot x2 else Slot x1
```

Comment:
We can actually implement this a little more cheaply without branching, something like this (haven't tested it):

```haskell
(Slot x1) `mappend` (Slot x2) =
    let !m = mask x1 maxBound
    in Slot $! (not m .&. x1) .|. (m .&. x2)
```